cpp/src/io/utilities/column_buffer.hpp

```diff
@@ -62,86 +62,184 @@ inline rmm::device_buffer create_data(data_type type,
 
+// forward declare friend functions
+template <typename string_policy>
+class column_buffer_base;
+
 /**
- * @brief Class for holding device memory buffers to column data that eventually
- * will be used to create a column.
+ * @brief Creates a column from an existing set of device memory buffers.
+ *
+ * @throws std::bad_alloc if device memory allocation fails
+ *
+ * @param buffer Column buffer descriptors
+ * @param schema_info Schema information for the column to write optionally.
+ * @param schema Optional schema used to control string to binary conversions.
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ *
+ * @return `std::unique_ptr<cudf::column>` Column from the existing device data
  */
-struct column_buffer {
-  column_buffer() = default;
-
-  // construct without a known size. call create() later to actually
-  // allocate memory
-  column_buffer(data_type _type, bool _is_nullable) : type(_type), is_nullable(_is_nullable) {}
-
-  // construct with a known size. allocates memory
-  column_buffer(data_type _type,
-                size_type _size,
-                bool _is_nullable,
-                rmm::cuda_stream_view stream,
-                rmm::mr::device_memory_resource* mr)
-    : column_buffer(_type, _is_nullable)
-  {
-    create(_size, stream, mr);
-  }
+template <class string_policy>
+std::unique_ptr<column> make_column(column_buffer_base<string_policy>& buffer,
+                                    column_name_info* schema_info,
+                                    std::optional<reader_column_schema> const& schema,
+                                    rmm::cuda_stream_view stream);
 
-  // move constructor
-  column_buffer(column_buffer&& col)            = default;
-  column_buffer& operator=(column_buffer&& col) = default;
+template <typename string_policy>
+class column_buffer_base {
+ public:
+  column_buffer_base() = default;
 
-  // copy constructor
-  column_buffer(column_buffer const& col)            = delete;
-  column_buffer& operator=(column_buffer const& col) = delete;
+  // construct without a known size. call create() later to actually allocate memory
+  column_buffer_base(data_type _type, bool _is_nullable) : type(_type), is_nullable(_is_nullable) {}
+
+  column_buffer_base(data_type _type,
+                     size_type _size,
+                     bool _is_nullable,
+                     rmm::cuda_stream_view stream,
+                     rmm::mr::device_memory_resource* mr)
+    : column_buffer_base(_type, _is_nullable)
+  {
+  }
 
   // instantiate a column of known type with a specified size.  Allows deferred creation for
   // preprocessing steps such as in the Parquet reader
-  void create(size_type _size, rmm::cuda_stream_view stream, rmm::mr::device_memory_resource* mr);
+  void create(size_type _size, rmm::cuda_stream_view stream, rmm::mr::device_memory_resource* mr)
+  {
+    size = _size;
+    _mr  = mr;
```

Comment:
> Shall we use a consistent naming convention to improve the code? I'm fine with either all data members with `_` prefix or all input with `_` prefix. The goal is to avoid such a mix use.

I agree, but the reason for the split right now is to distinguish between public and private data members.  I would like to make all the members private and prefixed, but that change had a huge ripple effect that I think should be addressed in a separate PR.

Reply:
> should be addressed in a separate PR.

works for me